optimization/univariate/un.metta

```diff
@@ -0,0 +1,87 @@
+;; ================================================================================
+;; Univariate Optimization (translated from moses/moses/optimization/univariate.cc)
+;; ================================================================================
+;;
+;; 
+;; - Inputs: Deme, Center instance, TruthTableBScore
+;; - Loop steps:
+;;   1. Estimate neighborhood size from Deme
+;;   2. Sample or generate neighbors near Center using sampleNewInstances
+;;   3. Score instances using transform with the complexity-based scorer
+;;   4. Pick best scored instance using returnBest folding
+;;   5. Set next Center to the best instance and iterate for bounded generations
+;; - Output: final center instance, updated Deme, and state tuple
+;;
+;; Simple theoretical example
+;; - Instance is a 1D knob with multiplicity three with values 0, 1, 2
+;; - Start center is 0; add neighbors 1 and 2
+;; - If score of 2 is better than 1, next center becomes 2
+;; - Repeat for limited generations and return best center with updated Deme
+;;
+;; Inputs
+;; - Deme contains Representation with knob map and tree, and an InstanceSet possibly scored
+;; - TruthTableBScore carries complexity coefficient and input table for scoring
+;; - Instance is the initial center from which neighborhood exploration starts
+;;
+;; Output
+;; - Returns an Instance, a Deme with scored instances, and a state tuple
+
+;; Entry point compatible with runMoses/expandDeme (same signature as hillClimbing)
+; (: univariate (-> Deme (TruthTableBScore $a) Instance (Instance Deme $state)))
+(= (univariate $deme $tTableBscorer $centerInst)
+   (let*
+       (
+         (() (println! (Entering univariate with center: $centerInst)))
+         ((mkTruthTableBScore $cpxCoeff $size $iTable) $tTableBscorer)
+         ((mkDeme $rep (mkSInstSet $instSet) $id) $deme)
+         ($discMap (getDiscMap $rep))
+         ($discMapLength (MultiMap.length $discMap))
+         ($maxGens (max 1 $discMapLength))
+         (() (println! (Max generations: $maxGens)))
+       )
+       (univariateLoop $deme $cpxCoeff $iTable $centerInst 0 $maxGens)
+   ))
+
+; (: univariateLoop (-> Deme Number (ITable $a) Instance Number Number (Instance Deme $state)))
+(= (univariateLoop (mkDeme $rep (mkSInstSet $instSet) $id)
+                    $cpxCoeff $iTable $centerInst $gen $maxGens)
+   (if (>= $gen $maxGens)
```

Comment:
This isn't the only termination criteria btw, you can take a look at the `termination_criterion` variable passed in the `eda/optimize.h` file lin number `125`.

Reply:
When you make the optimize function generic, not embedded only inside the univariate function, you need to pass that termination criteria as a function to it just like in the C++ version.